### PR TITLE
#payment_mode_configurations: Changes for add/edit/view mileage and per diem

### PR DESCRIPTION
**Pull request by yksolanki9**
For mileage and per-diem, we have to remove `PERSONAL_CORPORATE_CREDIT_CARD_ACCOUNT` from `allowed_payment_modes` and the default payment mode will always be `PERSONAL_ACCOUNT` irrespective of whether it is present or not in the `allowed_payment_modes` array

### CODE REVIEW

File: src/app/core/services/accounts.service.ts [hunk 1]
```diff
@@ -106,33 +106,50 @@ export class AccountsService {
...
   }
 
   //Filter and sort user accounts by allowed payment modes and return an observable of allowed accounts
   getAllowedAccounts(
     etxn: any,
     accounts: ExtendedAccount[],
-    isAdvanceEnabled: boolean,
-    isMultipleAdvanceEnabled: boolean
+    orgSettings: any,
+    expenseType: string
```
Comment (Julias0) [inline src/app/core/services/accounts.service.ts:117]: You can convert this to expenseType: 'MILEAGE' | 'PER_DIEM' | 'Something else'

If the expense types are well defined

File: src/app/core/services/accounts.service.ts [hunk 2]
```diff
@@ -106,33 +106,43 @@ export class AccountsService {
...
             paymentMode.acc.type = 'COMPANY_ACCOUNT';
             return allowedPaymentModes.some((allowedpaymentMode) => allowedpaymentMode === 'COMPANY_ACCOUNT');
           }
           return allowedPaymentModes.some((allowedpaymentMode) => allowedpaymentMode === paymentMode.acc.type);
         });
 
         const sortedPaymentModes = this.sortBasedOnAllowedPaymentModes(allowedPaymentModes, filteredPaymentModes);
-        const userPaymentModes = this.addMissingAccount(etxn, constructedPaymentModes, sortedPaymentModes);
+
+        const isMileageOrPerDiemExpense = ['MILEAGE', 'PER_DIEM'].includes(expenseType);
```
[the remark this answers is not in the document]
Comment (Julias0) [reply src/app/core/services/accounts.service.ts:146]: You can also do `expenseType: 'EXPENSE' | 'MILEAGE' | 'PER_DIEM'` in the arguments of this method, if this enum isn't used in many places

File: src/app/fyle/view-per-diem/view-per-diem.page.ts [hunk 3]
```diff
@@ -252,6 +256,17 @@ export class ViewPerDiemPage implements OnInit {
 
     this.updateFlag$.next(null);
 
+    if (this.view === ExpenseView.team) {
+      this.hidePaymentMode = false;
+    } else {
+      this.extendedPerDiem$
+        .pipe(
+          switchMap((extendedPerDiem) => this.accountsService.shouldPaymentModeBeHidden(extendedPerDiem)),
+          map((shouldPaymentModeBeHidden) => (this.hidePaymentMode = shouldPaymentModeBeHidden))
```
Comment (Julias0) [inline src/app/fyle/view-per-diem/view-per-diem.page.ts:265]: Assignments should happen inside subscribe

File: src/app/fyle/view-mileage/view-mileage.page.ts [hunk 4]
```diff
@@ -346,6 +350,17 @@ export class ViewMileagePage implements OnInit {
 
     this.updateFlag$.next(null);
 
+    if (this.view === ExpenseView.team) {
+      this.hidePaymentMode = false;
+    } else {
+      this.extendedMileage$
+        .pipe(
+          switchMap((extendedMileage) => this.accountsService.shouldPaymentModeBeHidden(extendedMileage)),
+          map((shouldPaymentModeBeHidden) => (this.hidePaymentMode = shouldPaymentModeBeHidden))
```
Comment (Julias0) [inline src/app/fyle/view-mileage/view-mileage.page.ts:359]: Same comment as below

**Review by Julias0 — CHANGES_REQUESTED** [review]
minor changes requested

File: src/app/core/services/accounts.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/core/services/accounts.service.ts:146]: Okay, will fix this in the other pr

File: src/app/fyle/view-per-diem/view-per-diem.page.ts (hunk 3 above)
Comment (yksolanki9) [reply src/app/fyle/view-per-diem/view-per-diem.page.ts:265]: Done

File: src/app/fyle/view-mileage/view-mileage.page.ts (hunk 4 above)
Comment (yksolanki9) [reply src/app/fyle/view-mileage/view-mileage.page.ts:359]: Done

**Review by Julias0 — APPROVED** [review]

File: src/app/core/services/accounts.service.ts [hunk 5]
```diff
@@ -106,22 +106,32 @@ export class AccountsService {
...
   //Filter user accounts by allowed payment modes and return an observable of allowed accounts
   getPaymentModes(
     accounts: ExtendedAccount[],
-    isMultipleAdvanceEnabled: boolean,
-    etxn?: any
+    orgSettings: any,
+    etxn?: any,
+    expenseType = 'EXPENSE'
   ): Observable<AccountOption[]> {
+    const isAdvanceEnabled = orgSettings?.advances?.enabled || orgSettings?.advance_requests?.enabled;
```
Comment (yksolanki9) [inline src/app/core/services/accounts.service.ts:119]: Not removing check for advances now, will need this for LD pr

**Review by gokulk16 — APPROVED** [review]

**Pull request merged**